src/Http/RequestFactory.php

```diff
@@ -60,25 +60,55 @@ public function setProxy($proxy)
 	 */
 	public function createHttpRequest()
 	{
+		$remoteAddr = !empty($_SERVER['REMOTE_ADDR']) ? $_SERVER['REMOTE_ADDR'] : NULL;
+		$remoteHost = !empty($_SERVER['REMOTE_HOST']) ? $_SERVER['REMOTE_HOST'] : NULL;
+
+		// use real client address and host if trusted proxy is used
+		$usingTrustedProxy = (bool)count(array_filter($this->proxies, function ($proxy) use ($remoteAddr) {
+			return $remoteAddr !== NULL && Helpers::ipMatch($remoteAddr, $proxy);
+		}));
+
 		// DETECTS URI, base path and script path of the request.
 		$url = new UrlScript;
-		$url->setScheme(!empty($_SERVER['HTTPS']) && strcasecmp($_SERVER['HTTPS'], 'off') ? 'https' : 'http');
+		$url->setScheme($scheme = $this->getScheme($usingTrustedProxy));
 		$url->setUser(isset($_SERVER['PHP_AUTH_USER']) ? $_SERVER['PHP_AUTH_USER'] : '');
 		$url->setPassword(isset($_SERVER['PHP_AUTH_PW']) ? $_SERVER['PHP_AUTH_PW'] : '');
 
-		// host & port
+		// host and port
+		$port = null;
 		if ((isset($_SERVER[$tmp = 'HTTP_HOST']) || isset($_SERVER[$tmp = 'SERVER_NAME']))
 			&& preg_match('#^([a-z0-9_.-]+|\[[a-f0-9:]+\])(:\d+)?\z#i', $_SERVER[$tmp], $pair)
 		) {
 			$url->setHost(strtolower($pair[1]));
 			if (isset($pair[2])) {
-				$url->setPort((int) substr($pair[2], 1));
+				$port = (int) substr($pair[2], 1);
 			} elseif (isset($_SERVER['SERVER_PORT'])) {
-				$url->setPort((int) $_SERVER['SERVER_PORT']);
+				$port = (int) $_SERVER['SERVER_PORT'];
+			}
+		} else if (!empty($_SERVER['SERVER_PORT'])) {
+			$port = (int) $_SERVER['SERVER_PORT'];
+		}
+
+		if ($usingTrustedProxy) {
+			if (!empty($_SERVER['HTTP_X_FORWARDED_PORT'])) {
+				$port = (int) $_SERVER['HTTP_X_FORWARDED_PORT'];
+			}
+
+			if (isset($_SERVER['HTTP_X_FORWARDED_FOR'])) {
+				$remoteAddr = trim(current(explode(',', $_SERVER['HTTP_X_FORWARDED_FOR'])));
+			}
+
+			if (isset($_SERVER['HTTP_X_FORWARDED_HOST'])) {
+				$remoteHost = trim(current(explode(',', $_SERVER['HTTP_X_FORWARDED_HOST'])));
 			}
 		}
 
-		// path & query
+		$isInvalidCombination = $scheme === 'http' && $port === 443 || $scheme === 'https' && $port === 80;
+		if ($port !== null && !$isInvalidCombination) {
+			$url->setPort($port);
+		}
```

Comment:
That is true, however in that case it doesn't really makes sense. I think that a lot of browsers would choose to do that anyway. But if you think that this shouldn't be there I can remove it - I just thought that it would be for the better.